### DISCUSSION TITLE
MacOS Development - July 2023

### DISCUSSION THREAD

**Discussion by scottfurry** [sqlitebrowser/sqlitebrowser#3410] [Show and tell]
Yeah,I'm just a little pleased with myself. Now to make the lives of others *easier*...

### Background
Reading through some of the latest bug reports, I noticed a trend in the reporting - reluctance to build the nightly in MacOS (at least, that's my *perception* anyways). Having gone through the process myself (and seriously, sometimes you have to stop beating your head against the wall from the *blood loss*), I think I understand what is happening, why it is a problem, and, most importantly, *how to work around the difficulties*.

I ran into all sorts of issues in trying to set up a development environment on MacOS Monterey. I have a 2014 Mac Mini I purchased recently at a vast discount(thank you, [Other World Computing (MacSales)](https://eshop.macsales.com/) - also with some great bits to expand/upgrade Macs). The Mac Mini works well. Setting up software pieces was its own challenge. And here we are...

It didn't help matters struggling to find an IDE that was both not antagonizing, as well as, helpful. I'm disappointed with Netbeans - the move to Apache Foundation broke the C++ environment. Despite what **_ANY_** web search tells you, after almost four years and eight releases, C++ building just **_DOES NOT WORK_**. VS Code and Eclipse are both not suitable C++ environments. They each have problems for C++ development and are, I find, frustrating to use. *If you're fighting with your IDE to do something simple and/or repetitive, you may have a bad IDE - or at least one that is not suitable for you.*

I'm resorting to [JetBrains CLion](https://www.jetbrains.com/clion/). It's functional, has a couple of workable quirks (of which I can rationalize why they exist), and is quite usable. Catch is cost - paying for software is something I haven't had to do for _decades_.

The major obstacle for DB4S nightly builds, I found, is the confluence of Qt and Homebrew. They have different priorities which are biting us in the back side.

### Qt
Qt is blasting away pushing out **Qt6**. The world would be a better place, in their belief, with their *latest-greatest* library.

Okay... but we're not ready. We have some dependency and platform issues to work through, yet. It was starting to feel like the *bad old days* where Gnome shoved gtk+3 down everyone's throats, causing massive disruption. Vast swaths of open source software was abandoned because people couldn't keep up with the constant barrage of API-breaking releases. Fortunately, Qt is not breaking things but their warp-speed advancement (they're on 6.5 branch!!!) is noted and not exactly helpful.

**Qt5** is currently the long term supported (or LTS) release. It is comforting that Qt is backporting some developments to the 5.15 branch. Overall, we're not being **forced** to change..._at the moment_..._for now_...

### Homebrew
Homebrew isn't exactly helping the cause but this is due to their operating practice.

Contrasting the situation - Debian, Fedora, and the like, work on the distribution level. These organizations ensure all the software pieces work together maintaining *bleeding-edge*, *testing (or next release)*, and *stable* distribution branches. Work on each branch involves not just integrating some software or library but also ensuring things, overall, work together.

Homebrew doesn't work like this - it is a *rolling-release* package distribution (think Arch Linux or similar). This leads to instances of *whack-a-mole* problems or *dependency-hell* situations. Any problems are quickly resolved but long-term maintenance of software is not exactly a priority.

#### Base Development Software
First, I am going on the assumption of a fresh OS install. Start from scratch. If you have Homebrew and Qt already installed - feel free to **_rip these pieces out_** in their entirety.

Uninstalling Qt is dependent on how Qt was installed. _Google-Fu_ your situation. There are just too many options for me to cover here.

For Homebrew - see the [FAQ answer here](https://docs.brew.sh/FAQ#how-do-i-uninstall-homebrew). Homebrew utilizes an *uninstall* script to clean things up. It is quite robust and thorough.

Ultimately, there is the **_Nuclear Option_** - re-install MacOS. The *cost-benefit analysis* of hunting down some *make-it-work* fix versus setting up the OS for a *clean environment* is an exercise best left to the reader. I was able to create a **install USB drive** in order to start with a fresh MacOS Monterey(v12.6.7) without having to go through the painful process of updating. YMMV. Caveat Emptor.

#### Apple
Homebrew will demand XCode be installed. Do this now. Install both `XCode` and the `XCode Command Line Tools`. Homebrew install will ensure everything is *correct* before continuing. 

#### Homebrew
My guiding principle WRT Homebrew was to minimize **ANY**/**ALL** dependencies - period!!!!
Initially, I only installed some binary-only casks (VLC, Gimp, Inkscape or similar) that **did not** have dependencies.

For development software, I pulled in the following:

`brew install cmake clang-format llvm meson nano ninja qt@5 shellcheck sqlite`

This pulls in some important pieces (i.e. `autoconf`, `glib`, `python`, `readline`, etc). However, the dependencies are *known* and *manageable*. No surprises or cruft. We **cannot** use the shortened `qt` with homebrew as that will pull-in **Qt6**. Again, Homebrew is *latest-greatest* and not what we want at the moment. Use `qt@5` to **explicitly** install the **Qt5** branch.
 
#### Homebrew Qt@5
Because of Homebrew's setup, CLion was able to pick up the location of **Qt5** _somehow_. However, if building from the command-line, you will need to execute the following **ANYTIME** you start to build and need to reference Homebrew's **qt@5**:
```
export PATH="$(brew --prefix qt@5)/bin:$PATH"
export LDFLAGS="-L$(brew --prefix qt@5)/lib"
export CPPFLAGS="-I$(brew --prefix qt@5)/include"
export PKG_CONFIG_PATH="$(brew --prefix qt@5)/lib/pkgconfig"
```
The instructions are derived from Homebrew's information printed to the console after **qt@5** installation.

#### Homebrew SQLite
Unfortunately, we have to tweak the build to direct usage of Homebrew's SQLite (and **NOT** version installed by XCode). And this was the *magic-sauce* I discovered this week.

CMake can be *guided* to other locations, either through the command line (prepend directive with `-D`)  or via the IDE, by populating CMake variables. These changes make integration with CMake more understandable.

The [`CMAKE_INCLUDE_PATH`](https://cmake.org/cmake/help/latest/variable/CMAKE_INCLUDE_PATH.html#variable:CMAKE_INCLUDE_PATH) and [`CMAKE_LIBRARY_PATH`](https://cmake.org/cmake/help/latest/variable/CMAKE_LIBRARY_PATH.html#variable:CMAKE_LIBRARY_PATH) variables are **semi-colon** lists of strings containing directory locations for CMake to search when generating software. To direct CMake to use the Homebrew installed SQLite we need to give the following either on the command-line or in the IDE:
```shell
-DCMAKE_INCLUDE_PATH="/usr/local/opt/sqlite/include"
-DCMAKE_LIBRARY_PATH="/usr/local/opt/sqlite/lib"
```
Again, the `-D` is prepended to variable for command-line/IDE usage. Also, for **M1 Macs** the path value to SQLite **WILL BE** different. Use `$(brew --prefix sqlite)` on the command line to retrieve the exact value. Your IDE may not be *flexible* in allowing to *script* the path. Check IDE documentation.

If all goes correctly, CMake should respond with something similar to...
```
-- The C compiler identification is Clang 16.0.6
-- The CXX compiler identification is Clang 16.0.6
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/local/opt/llvm/bin/clang - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/local/opt/llvm/bin/clang++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found SQLite3: /usr/local/opt/sqlite/include (found version "3.42.0") 
-- Configuring done (8.7s)
-- Generating done (0.4s)
-- Build files have been written to: /[path to DB4S project root]/build
```
Building should pull in the homebrew-installed SQLite library. You can confirm with the DB4S `About Dialog`:
![DB4S_About_2023Jul](https://github.com/sqlitebrowser/sqlitebrowser/assets/6686571/d7160ee6-a71e-4a15-b793-c8e2088138fe)

And you have a *shiny*, *new* DB4S app to work from. I celebrated with a coffee. Your celebratory beverage-of-choice may differ. Exercise caution.

### Local QScintilla
This was painful. What I had to *untangle* was *how exactly* were we pulling in QScintilla in the **_first place_**. It didn't help that the directory structure of the library has changed between 2.13 and 2.14. However, I got there...

What I realized was _how QScintilla was changed_ previously. In the project tree, under `[project_root]/libs/` were the files `DB4S_PATCH_05` and `DB4S_PATCH_06`. These files documented the changes to QScintilla both to minimize installed components but to also modify the qmake build initially to generate a static library (i.e. `libqscintilla2.a`). We have since generated a custom `CMakeList.txt` file to govern the build for QScintilla for packaging. But what if someone whats to use local...?

Download the latest copy of [QScintilla](https://www.riverbankcomputing.com/software/qscintilla/download) from upstream. For convenience of explanation, I saved it in the same directory holding the local copy of DB4S project from GitHub. Open Finder and navigate to this directory, location of the saved download.

Neat trick is to enable Finder's *Path Bar* (shows up at bottom of Finder window and shows the full path to what is shown/selected. Enable this feature with Finder Menu Bar -> `View` -> `Show Path Bar`. Now, right-click the directory where download was saved in the path bar and select "New Terminal at Folder"
![Path_Bar_Shortcut](https://github.com/sqlitebrowser/sqlitebrowser/assets/6686571/a881e16f-6072-4a05-ba87-31a2f004bde8). A terminal window will open with the current working directory already set to the download location.

Now for some _command-line-fu_. **_FIRST_** ensure you have setup environment to use **qt@5** by executing all the export statements shown above.

Next, we need to establish install location for QScintilla. Execute:
```
mkdir /opt/local
sudo chown -fR [user name]:admin /opt/local
sudo chmod 775 /opt/local
```
For the `sudo` commands, enter your Mac password. I'm pulling a page from Homebrew here. First, we **do not want** to pollute `/usr/local`. So, we'll just avoid that location and opt instead for the safer `/opt/local` directory (unless you're using MacPorts, then you're on your own here). Second, we establish permissions and ownership on the directory `/opt/local` similar to how Homebrew uses permissions/ownership for it's install. We're just copying the methodology for a different location.

Now, let's build QScintilla. Extract the sources (either `tar xf QScintilla_src-X.Y.Z.tar.gz` or `unzip QScintilla_src-X.Y.Z.zip` depending on what archive was downloaded). Change into the extracted directory(`cd QScintilla_src-X.Y.Z`).

We need to modify the existing sources:
```diff
diff --git a/src/qscintilla.pro b/src/qscintilla.pro
index c6f143a..509e2dc 100644
--- a/src/qscintilla.pro
+++ b/src/qscintilla.pro
@@ -21,21 +21,8 @@
 !win32:VERSION = 15.2.1
 
 TEMPLATE = lib
-CONFIG += qt warn_off thread exceptions hide_symbols
-
-CONFIG(debug, debug|release) {
-    mac: {
-        TARGET = qscintilla2_qt$${QT_MAJOR_VERSION}_debug
-    } else {
-        win32: {
-            TARGET = qscintilla2_qt$${QT_MAJOR_VERSION}d
-        } else {
-            TARGET = qscintilla2_qt$${QT_MAJOR_VERSION}
-        }
-    }
-} else {
-    TARGET = qscintilla2_qt$${QT_MAJOR_VERSION}
-}
+TARGET = qscintilla2
+CONFIG += qt warn_off thread exceptions hide_symbols staticlib debug_and_release
 
 macx:!CONFIG(staticlib) {
     QMAKE_POST_LINK += install_name_tool -id @rpath/$(TARGET1) $(TARGET)
@@ -43,6 +30,8 @@ macx:!CONFIG(staticlib) {
 
 INCLUDEPATH += . ../scintilla/include ../scintilla/lexlib ../scintilla/src
 
+QMAKE_CXXFLAGS += -std=c++11
+
 !CONFIG(staticlib) {
     DEFINES += QSCINTILLA_MAKE_DLL
 
@@ -61,38 +50,21 @@ macx:lessThan(QT_MAJOR_VERSION, 6) {
 # Work around QTBUG-39300.
 CONFIG -= android_install
 
-# For old versions of GCC.
-unix:!macx {
-    CONFIG += c++11
-}
-
 # Comment this in if you want the internal Scintilla classes to be placed in a
 # Scintilla namespace rather than pollute the global namespace.
 #DEFINES += SCI_NAMESPACE
 
-target.path = $$[QT_INSTALL_LIBS]
+target.path = $$[PREFIX]/lib
 INSTALLS += target
 
-header.path = $$[QT_INSTALL_HEADERS]
+header.path = $$[PREFIX]/include
 header.files = Qsci
 INSTALLS += header
 
-trans.path = $$[QT_INSTALL_TRANSLATIONS]
+trans.path = $$[PREFIX]/translation
 trans.files = qscintilla_*.qm
 INSTALLS += trans
 
-qsci.path = $$[QT_INSTALL_DATA]
-qsci.files = ../qsci
-INSTALLS += qsci
-
-features.path = $$[QT_HOST_DATA]/mkspecs/features
-CONFIG(staticlib) {
-    features.files = $$PWD/features_staticlib/qscintilla2.prf
-} else {
-    features.files = $$PWD/features/qscintilla2.prf
-}
-INSTALLS += features
-
 HEADERS = \
     ./Qsci/qsciglobal.h \
     ./Qsci/qsciscintilla.h \
```

As described above, we have patch files in the source tree from previous iteration. We need the diff shown above to accommodate upstream updates. Take the changes given and save this to the file `qsci.diff` in the same path as where QScintilla archive was downloaded.

We can go one of two ways,...
a) directly - type into the command-line, with the terminal CWD pointing to extracted QScintilla sources,:
```
patch -p1 <../qsci.diff
```

b) with git - again, CWD is in extracted QScintilla. Enter:
```
git init
git add .
git commit -m "Baseline"
git apply ../qsci.diff
```
This method has the added benefit of being able to *blow away* any changes/problems you encounter. You can revert with the commands:
```
git clean -dfx
git reset --hard HEAD
```
to get back to the _"Baseline"_, or what was downloaded.

With terminal setup and sources patched, we now build:
```
% cd src
% qmake qscintilla.pro
# qmake response...
% make
# ... and go for a coffee ...
# output:  release/moc_ScintillaQt.o
# /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib -s libqscintilla2.a
% make INSTALL_ROOT="/opt/local/QScintilla" install
```

And confirming the output with a directory listing:
```
 % ls -al /opt/local/QScintilla
total 24
drwxr-xr-x  6 [user name]  admin   192 17 Jul 12:10 .
drwxrwxr-x  4 [user name]  admin   128 17 Jul 12:10 ..
-rw-r--r--@ 1 [user name]  admin  8196 17 Jul 12:10 .DS_Store
drwxr-xr-x  4 [user name]  admin   128 17 Jul 12:10 include
drwxr-xr-x  3 [user name]  admin    96 17 Jul 19:22 lib
drwxr-xr-x  7 [user name]  admin   224 17 Jul 19:22 translation
```

Okay...move over to building DB4S. Either with the terminal or in your IDE, enter the following:
```
-DFORCE_INTERNAL_QSCINTILLA=OFF
-DCMAKE_INCLUDE_PATH="[path to sqlite from above];/opt/local/QScintilla/Qsci/include"
-DCMAKE_LIBRARY_PATH="[path to sqlite from above];/opt/local/QScintilla/lib"
# ... other CMake directives as desired ...
```

Letting CMake do its thing gets me the following:
```shell
-- The C compiler identification is Clang 16.0.6
-- The CXX compiler identification is Clang 16.0.6
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/local/opt/llvm/bin/clang - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/local/opt/llvm/bin/clang++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found QScintilla: /opt/local/QScintilla/lib/libqscintilla2.a (found suitable version "2.14.1", minimum required is "2.8.10") 
-- Found SQLite3: /usr/local/opt/sqlite/include (found version "3.42.0") 
-- Configuring done (11.3s)
-- Generating done (0.3s)
-- Build files have been written to: /[path to DB4S project root]/build
```

At this point, I would be _rushing_ to talk about how everything built _JUST FINE_ but ran into a snag. I know exactly how it happened and why. Safe to say, we have some clean up to do in our `FindQScintilla.cmake` module.

The keen observer will note something off about the `CMAKE_INCLUDE_PATH` given above. This is the full path to where the `qscintilla.h` file can be found following the build instructions.

If QScintilla is not installed as part of **qt@5** infrastructure **_AND_** the full path to `qscintilla.h` is not _explicitly_ set, CMake will fail. With the correct path set, CMake is obliviously happy however, the build will fail complaining about not finding `Qsci/qscintilla.h`.

It seems our custom `FindQScintilla.cmake` module only modifies **Qt5** directories with a regex command (see `[project_root]/cmake/FindQScintilla.cmake`:line 50). Our cmake module **is assuming** QScintilla is located in the **Qt5** structure. I do not think this is just a *Mac* or *Unix* problem. We've just been really lucky to date not having found this bug till now.

I got a fair distance here. Just didn't hit the *endzone* unfortunately. I'll come back to this as time permits and/or things change.

**Comment by justinclift** [comment]
> ... reluctance to build the nightly in MacOS.

Errr, what?

---

On a related topic, I'm very overdue to give remote access to @lucydodo to the newly set up M1 mac mini here.  Um... give me a few more days on that.  It shouldn't be too hard to do, I just need to focus on that for a bit. :smile:

When that's done and the M1 Mac is also building macOS binaries on every new commit then the nightly builds may not be needed any more.

**Reply by lucydodo** [reply]
There's no rush, take your time.

**Comment by scottfurry** [comment]
> > ... reluctance to build the nightly in MacOS.
> 
> Errr, what?

edited - (at least, that's my *perception* anyways)

**Reply by justinclift** [reply]
Oops.  Wish I'd understood you were trying and struggling with it.  I'd gone through the setup process for our nightly builds on macOS several times, so I could have documented it.  <-- Clearly that should be "should have documented it". :man_facepalming:

**Reply by scottfurry** [reply]
That's the problem with our current *bottom-line* focused society - what *should be done* isn't what is done (insert Jeff Goldblum/Jurasic Park gif here).

And it isn't always cost-effective for some to do documentation because of the *break-everything* thinking that is prevalent in current software ecosystem. Something will change making documentation irrelevant. It's a vicious cycle.

Sad we are so far removed from the heady days of Unix and the "do one thing well" idealism.

But that's me on a soap box again...
Shocking, I know. 😉

**Comment by scottfurry** [comment]
I hit a road block with the build but I've documented as much as possible. We have an issue in a CMake module to clean up, I believe. There's some assumptions that tripped me up today.

I can live with what is documented, so far.

**Comment by lucydodo** [comment]
**Qt**: As far as I know, official support for Qt 5 has ended, so I've been trying play around with migrating to Qt 6,
but it's not as easy as I'd like it to be. I'll share it with the team when I get a little more clarity.

**Homebrew**: When using Homebrew, dependency issues can be resolved by explicitly specifying version.
(which will definitely help you deploy more stable builds)

Additionally, the SQLite package is deployed for the build using its [own tab](https://github.com/sqlitebrowser/homebrew-sqlite3), and CMake is set up based on that package.
If you install that package, you shouldn't need the next two options you mentioned.
```
-DCMAKE_INCLUDE_PATH="/usr/local/opt/sqlite/include"
-DCMAKE_LIBRARY_PATH="/usr/local/opt/sqlite/lib"
```

Anyway, my personal opinion is that the documentation is a bit lacking for the macOS build.
I have a plan for writing some new documentation on that (within the next week at the least). Let me know if you have any feedback. :)

**Reply by scottfurry** [reply]
> **Qt**: As far as I know, official support for Qt 5 has ended, so I've been trying play around with migrating to Qt 6, but it's not as easy as I'd like it to be. I'll share it with the team when I get a little more clarity.

@lucydodo - No. See [this Qt blog post](https://www.qt.io/blog/qt-5.15-extended-support-for-subscription-license-holders) indicating that Qt5.15 branch is LTS and will be supported until mid-2025.

Certainly the pressure is not off on us. There are organizations releasing software dependent on Qt who are **_rushing_** to use the latest-greatest. This isn't making our lives any easier - which is what software development should **NOT BE LIKE**. Generating grey hair in others is always bad.

**Reply by lucydodo** [reply]
>This extension of the LTS is available for those with subscription licenses.

I believe this is only true if we have a subscription to a commericial license.

**Reply by scottfurry** [reply]
> **Homebrew**: When using Homebrew, dependency issues can be resolved by explicitly specifying version. (which will definitely help you deploy more stable builds)
> 
> Additionally, the SQLite package is deployed for the build using its [own tab](https://github.com/sqlitebrowser/homebrew-sqlite3), and CMake is set up based on that package. If you install that package, you shouldn't need the next two options you mentioned.
> 
> ```
> -DCMAKE_INCLUDE_PATH="/usr/local/opt/sqlite/include"
> -DCMAKE_LIBRARY_PATH="/usr/local/opt/sqlite/lib"
> ```

@lucydodo - Homebrew does allow one to `pin` a version or install from a personal `tap`. However, what you are suggesting is not exactly correct. Homebrew installed sqlite as `keg-only` so as not to conflict with XCode installed SQLite. This should be obvious when CMake is executed when the version and path to SQLite is printed out. To use the homebrew SQLite, one has to tell CMake where to find it. What is being suggested sounds like _breakage_ or _forcing_ homebrew installed on MacOS. This situation is very much **NOT RECOMMENDED** for general usage.

It is one thing to seek to _break an operating system_ with everyday usage. It is entirely another matter to do so because one doesn't agree with how the pieces work together or how they were installed by forcing will upon the system that makes the computer operating system unstable. IMHO, that's not a mature code mentality, but that's just my $|£|€|¥ 0.02 worth.

**Reply by lucydodo** [reply]
We can also set `keg_only` for our SQLite package. Would that be better? :)

**Reply by lucydodo** [reply]
or user can link sqlite package on demand with the `brew link`, or `brew unlink` command.

**Reply by scottfurry** [reply]
post output of `brew info sqlite` of platform you are referencing.

**Reply by lucydodo** [reply]
Oh, I'm talking about the `sqlitefts5` package. :)

**Reply by scottfurry** [reply]
```
% brew info sqlite
==> sqlite: stable 3.42.0 (bottled) [keg-only]
Command-line interface for SQLite
https://sqlite.org/index.html
/usr/local/Cellar/sqlite/3.42.0 (11 files, 4.5MB)
  Poured from bottle using the formulae.brew.sh API on 2023-07-15 at 02:55:23
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/sqlite.rb
License: blessing
==> Dependencies
Required: readline ✔
==> Caveats
sqlite is keg-only, which means it was not symlinked into /usr/local,
because macOS already provides this software and installing another version in
parallel can cause all kinds of trouble.

If you need to have sqlite first in your PATH, run:
  echo 'export PATH="/usr/local/opt/sqlite/bin:$PATH"' >> ~/.zshrc

For compilers to find sqlite you may need to set:
  export LDFLAGS="-L/usr/local/opt/sqlite/lib"
  export CPPFLAGS="-I/usr/local/opt/sqlite/include"

For pkg-config to find sqlite you may need to set:
  export PKG_CONFIG_PATH="/usr/local/opt/sqlite/lib/pkgconfig"
==> Analytics
install: 952 (30 days), 952 (90 days), 952 (365 days)
install-on-request: 109 (30 days), 109 (90 days), 109 (365 days)
build-error: 0 (30 days)
```

**Reply by scottfurry** [reply]
Homebrew does not list a `sqlitefts5` formula. I don't see an install for the `FTS5` extension in Homebrew. I could be wrong.

**Reply by lucydodo** [reply]
Oh, I didn't explain myself well enough.
The `sqlitefts5` package is in our repository (and not set to`key-only`)

You can import our repository and install the package with the following command.
```bash
$ brew tap sqlitebrowser/sqlite3
$ brew install sqlitefts5
```

**Reply by scottfurry** [reply]
:scream: :scream: :scream:
_We have a homebrew tap!?!?!_
:scream: :scream: :scream:

That's what I get for not reading documentation.

**Reply by justinclift** [reply]
Heh Heh Heh

https://github.com/sqlitebrowser/sqlitebrowser/blob/5d6ce7fb5b167c469f1a3c263d9e54109cfcc3e3/installer/macos/build_sqlitebrowser_nightly-intel.sh#L75

**Reply by scottfurry** [reply]
The words **_undocumented feature_** comes to mind for some reason...

**Reply by justinclift** [reply]
That Homebrew tap also automatically updates to the newest version of SQLite as well. :smile:

**Reply by scottfurry** [reply]
Which then begs the questions...
- Are users advised of this action?
- Are they made aware of the homebrew install?
- Have measures been taken to not break the user's install?

I'm looking at this whole script and left with the impression that the user's environment may not be _respected_ by the install.

**Reply by justinclift** [reply]
Er... that Homebrew install is only needed for the build process.  It's nothing to do with end users. :smile:

**Reply by justinclift** [reply]
Ahhh.  I think there's a misunderstanding.  That tap is only for the build process to find our customised version of SQLite (with extra extensions, etc).

Mid-way through the build process, the nightly build script runs `macdeployqt`:

https://github.com/sqlitebrowser/sqlitebrowser/blob/5d6ce7fb5b167c469f1a3c263d9e54109cfcc3e3/installer/macos/build_sqlitebrowser_nightly-intel.sh#L106

That copies the 3rd party libraries into the DB4S .app (eg SQLite, Qt, etc).  That way the .app has them at hand, and users don't need to install the 3rd party libraries themselves.

Hopefully that makes sense?

**Reply by scottfurry** [reply]
Deployment script. Got it. :+1:
